util/addon.c

```diff
@@ -0,0 +1,47 @@
+#include <stdlib.h>
+#include <wayland-server-core.h>
+
+#include "wlr/util/addon.h"
+
+void wlr_addon_set_init(struct wlr_addon_set *set) {
+	wl_list_init(&set->addons);
+}
+
+void wlr_addon_set_finish(struct wlr_addon_set *set) {
+	struct wlr_addon *addon, *tmp;
+	wl_list_for_each_safe(addon, tmp, &set->addons, link) {
+		wlr_addon_destroy(addon);
+	}
+}
+
+struct wlr_addon *wlr_addon_create(struct wlr_addon_set *set, void *data,
+		void *owner, wlr_addon_finish_data_func_t finish_data) {
+	struct wlr_addon *addon = calloc(1, sizeof(*addon));
```

Comment:
Instead of allocating a `wlr_addon` here, can we make the caller responsible for embedding it in one of its struct?

Reply:
```c
struct wlr_xdg_surface_state {
    struct wlr_addon addon;
}
```
Like that?

Reply:
Yup. Doesn't necessarily need to be the first field, and users can get back the parent struct with something like `struct wlr_xdg_surface_state *xdg_state = wl_container_of(addon, xdg_state, addon)`.

Reply:
Yeah, that would be better - then instead of creating, one would do `wlr_addon_set_insert(&surface->addon_set, &xdg_surface->addon)`.

Reply:
Actually, if we want to hide the implementation, we don't want to expose `wlr_addon` with its `wl_list link` at all.

Reply:
We can mark the `wl_list` field as private state:

```c
struct wlr_addon {
    // private state
    struct wl_list link;
};
```